django/contrib/auth/hashers.py

```diff
@@ -416,26 +446,36 @@ def encode(self, password, salt):
         data = bcrypt.hashpw(password, salt)
         return "%s$%s" % (self.algorithm, data.decode('ascii'))
 
+    def decode(self, encoded):
+        algorithm, empty, algostr, work_factor, data = encoded.split('$', 4)
+        assert algorithm == self.algorithm
+        salt, checksum = data[:22], data[22:]
+        return {
+            'algorithm': algorithm,
+            'algostr': algostr,
+            'work_factor': work_factor,
+            'salt': salt,
+            'checksum': checksum,
+        }
+
     def verify(self, password, encoded):
         algorithm, data = encoded.split('$', 1)
         assert algorithm == self.algorithm
         encoded_2 = self.encode(password, data.encode('ascii'))
         return constant_time_compare(encoded, encoded_2)
 
     def safe_summary(self, encoded):
-        algorithm, empty, algostr, work_factor, data = encoded.split('$', 4)
-        assert algorithm == self.algorithm
-        salt, checksum = data[:22], data[22:]
+        decoded_entry = self.decode(encoded)
         return {
-            _('algorithm'): algorithm,
-            _('work factor'): work_factor,
-            _('salt'): mask_hash(salt),
-            _('checksum'): mask_hash(checksum),
+            _('algorithm'): decoded_entry['algorithm'],
+            _('work factor'): decoded_entry['work_factor'],
+            _('salt'): mask_hash(decoded_entry['salt']),
+            _('checksum'): mask_hash(decoded_entry['checksum']),
         }
 
     def must_update(self, encoded):
-        algorithm, empty, algostr, rounds, data = encoded.split('$', 4)
-        return int(rounds) != self.rounds
+        decoded_entry = self.decode(encoded)
+        return int(decoded_entry['work_factor']) != self.rounds
```

Comment:
This `int` can probably be dropped too?

Reply:
`work_factor` is still a string. I can make the underlying data type an int if you like though.